proto/tendermint/types/types.proto

```diff
@@ -151,7 +213,38 @@ message BlockMeta {
 
 // TxProof represents a Merkle proof of the presence of a transaction in the Merkle tree.
 message TxProof {
-  bytes                   root_hash = 1;
-  bytes                   data      = 2;
-  tendermint.crypto.Proof proof     = 3;
+  repeated bytes    row_roots = 1;
+  repeated bytes    data      = 2;
+  repeated NMTProof proofs    = 3;
+}
+
+// MalleatedTx wraps a transaction that was derived from a different original
+// transaction. This allows for tendermint to track malleated and original
+// transactions
+message MalleatedTx {
+  bytes original_tx_hash = 1;
+  bytes tx               = 2;
+  uint32 share_index     = 3;
+}
+
+// Proof represents proof of a namespace.ID in an NMT.
+// In case this proof proves the absence of a namespace.ID
+// in a tree it also contains the leaf hashes of the range
+// where that namespace would be.
+message NMTProof {
+  // start index of this proof.
+  int32 start = 1;
+  // end index of this proof.
+  int32 end = 2;
+  // Nodes that together with the corresponding leaf values can be used to
+  // recompute the root and verify this proof. Nodes should consist of the max
+  // and min namespaces along with the actual hash, resulting in each being 48
+  // bytes each
+  repeated bytes nodes = 3;
+  // leafHash are nil if the namespace is present in the NMT. In case the
+  // namespace to be proved is in the min/max range of the tree but absent, this
+  // will contain the leaf hash necessary to verify the proof of absence. Leaf
+  // hashes should consist of the namespace along with the actual hash,
+  // resulting 40 bytes total.
```

Comment:
In that case, wouldn't if be better to have, instead of bytes, a structure holding the namespace and the hash?
"better" in terms of readability and maintainability